### ISSUE TITLE
[Feature Request] Sublime/VSC Style Minimap 

### ISSUE THREAD

**Issue by shaunsingh** [Kethku/neovide#669]
Hi, I was wondering if it was possible to have a Sublime style minimap in the gui?

For example, the akiyosi fork of goneovim has this on the right hand side : 
![image](https://user-images.githubusercontent.com/71196912/119147720-d70f8780-ba19-11eb-9eac-b2bec9b39c70.png)

There is a plugin for neovim that accomplishes something similar (minimap.vim), but its limited to the text rendering of neovim so it uses braille font (which doesn't look as good). 

This may not be possible right now (I assume it would require multigrid support as you need to draw the window over neovim), but is this something you can look into for the future?

Sincerely 
Shaurya

**Comment by PyGamer0** [comment]
[dup of this issue](https://github.com/Kethku/neovide/issues/429), but its closed. But even I agree neovide should have a custom minimap like goneovim with syntax highlighting.

**Comment by shaunsingh** [comment]
the issue with making it a plugin is that you can't take advantage of gui features, and so you would be either 
1. limiting to display uncolored monospaced text 
2. making an implementation that can "display" an image using cat and put it in a floating window 
The first one doesn't look very good, and the second one doesn't have an implementation yet, but even if it did I imagine it would be slow

**Comment by j4qfrost** [comment]
Wouldn't be too complicated for the view, interacting might be a bit
tricky. I think we should approach the problem with a plugin solution so
that we can get some code reuse. I'll free up some time next week to look
into this.

**Comment by j4qfrost** [comment]
@shaunsignh just saw your comment. When I say plugin I mean that neovide
would have its own plugin system. We would be able to leverage internals
and have more control over functionality.

**Comment by shaunsingh** [comment]
Ah, sorry I misinterpreted it that. 

A plugins system sounds like a very good idea. Then you could add new features without decreasing performance. A plugin system could also open up the door to other features e.g. custom tablines or markdown previews

**Comment by shaunsingh** [comment]
I was looking at goneovim's implementation here: https://github.com/akiyosi/goneovim/blob/master/editor/minimap.go, and it looks like they just spawn a second incredibly small/long neovim process. Is this the best way to implement it?

**Comment by Kethku** [comment]
I'm definitely interested in peoples thoughts here, but I have a vision for eventually exposing a graphics api to neovim itself. Imagine a series of functions which are loaded on startup that let a third party neovim plugin add graphics to a given buffer. This way anybody could add arbitrary graphical features to the neovim windows from whatever plugin language they like.

I mentioned the idea in the previous minimap issue and I think it has merit and imho would make neovide way more compelling to use as a gui

**Comment by Kethku** [comment]
Imagine live coding a procedural vector graphics animation into the window you're coding on. Thats the sort of thing I'd like to enable

**Comment by Kethku** [comment]
> 
> 
> Ah, sorry I misinterpreted it that.
> 
> A plugins system sounds like a very good idea. Then you could add new features without decreasing performance. A plugin system could also open up the door to other features e.g. custom tablines or markdown previews

Just to respond here. Neovim plugins can also be remote processes which call functions asynchronously. Exposing graphics functions to neovim plugins doesn't necessarily slow things down

**Comment by shaunsingh** [comment]
That sounds awesome! One of the biggest things I miss about emacs is the inline LaTeX previews and being able to insert images, you can obviously open up a live preview in a browser for neovim, but having everything in one place makes it much easier

**Comment by Kethku** [comment]
Its a pretty big design problem because we will have to expose a remote api which is easy enough to understand while still handling all the use cases, but using a minimap plugin as a target use case seems like a great way to go. It could help motivate what needs exposed.

This sort of graphics api is high on my list for what to work on after the current set of pinned issues, but if somebody wants to take a stab at it sooner, I'd love to work with them. Warning though, I'm going to be pretty particular and will likely want a fair bit of iteration on what it looks like because I see it as a huge potential draw for neovide in general

**Comment by Kethku** [comment]
Which is not to say don't go for it :P just that I'm gonna want to work closely with anybody who does :)

**Comment by fredizzimo** [comment]
Just some random ideas for the future if someone starts to work on this.

Skia has LUA bindings https://chromium.googlesource.com/skia/+/chrome/m49/site/user/special/lua.md, so that could be one way of supporting a graphics API. If we are willing to run these graphical plugins on the Neovide side. However, that's not very user friendly for plugin writers, since they would have to use RPC calls to communicate with the plugin running on the Nvim side. But on the other hand this would allow for great performance with animations and stuff, and also very powerful due to Skia.

Another potential alternative would be if we could take that API, and using some code generator automatically convert the API to RPC calls. But that will likely be too slow, even if we added batching, since the API is not designed for remote usage and skia is complex

There are also many other more minimal lua graphcis APIs that we could mimic.

Another completely different approach is something more declarative, like HTML, and modern CSS with animations (but without javascript). It would obviously probably not be HTML and CSS, just something that slightly resembles it(otherwise we are reinventing Electron...). This would make it easy to use by plugin writers, while still maintaining a lot of flexibility.

**Comment by Kethku** [comment]
A key component IMHO is that whatever solution should be easily integratable with standard neovim config. The goal (with all of neovide up till now) is to feel as neovim native as possible while still enabling new things. Maybe I'm just fussy about it, but to me that sorta leans towards hand written or generated api exposed via Lua rather than an external Lua process or html like language.

Also re performance, I think it depends on what the end goal is. Something like minimap doesn't have to be super fast to get something reasonable and useful. In which case an rpc solution may be good enough and give lots of api flexibility. However it's possible once something like this is available we will immediately reach for it and be disappointed by how slow it is. 

Idk for sure though! And I would love to chat more about it. Any solution is better than no solution cause we can iterate on it if it doesn't feel right.

**Comment by fredizzimo** [comment]
Yes, I was just throwing in some random ideas. it obviously needs a lot more thought before implementing.

But just a small note, I was never talking about using an external process, but there's still two different places where we can run the Lua scripts. They can can be inside the Neovide process where they could use the Skia API directly but needs to use RPC to communicate with the rest of the Neovim ecosystem and possible the other part of the plugin inside Neovim. Or they could be inside the Neovim process, in which case all the drawing commands needs to be done through some kind of RPC interface.

**Comment by AleksanderBobinski** [comment]
There exists a plugin - https://github.com/Isrothy/neominimap.nvim. Has anyone tried it with neovide?
